### ISSUE TITLE
Cross Compiling for Freescale i.MX6Solo (ARM, little endian)

### ISSUE THREAD

**Issue by kidahl** [mikebrady/shairport-sync#325]
Could  someone point me in the right direction to cross compile the shairport-sync to a Freescale ARM unit?

Long story: I was a Kickstarter backer for Olive One, a great AMP with an embedded linux server. Olive has now stopped answering requests putting the unit with all the backers in limbo. 
I have the root password, but the unit does not provide a compiler. The Linux environment seems plain enough (500MB RAM, plenty disk, rw filesystem etc), I just need to transfer the binaries to the device. However, I cannot find any ARM binaries anywhere.

I just built the entire OpenWRT distro with the shairport plugin go get at ARM binaries, but these seem to have a custom binary format (.ipk).

Is it possible to build a statically linked ARM binary from an OSX or Debian machine? If so, I need some pointers, as cross-compiling seems to be quite a messy affair.

A lot of backers depend on this. Thanks.

**Comment by mikebrady** [comment]
Gosh, I don't have a line on that at all. Do you know anything about their software? Are they using some existing open source stuff?

**Comment by kidahl** [comment]
We know very little, Olive just stopped responding after promising releases for a long time. But if someone could provide me with binaries from ARM- builds, I could give them a try. For example the raspi build.

**Comment by joerg-krause** [comment]
Please, have a look at the [Buildroot](https://buildroot.org) and [Yocto](https://www.yoctoproject.org) project. Buildroot is quite easy to use...

**Comment by mikebrady** [comment]
@kidahl, if you send me a mail at my email address, I can send you Raspberry Pi binaries.

**Comment by kidahl** [comment]
@mikebrady 
Thanks Mike, it´s worth a try. We are over 1400 backers that are pretty nervous at this point. You can reach ma at kidahl (at) gmail.com

I have since discovered that the unit runs libc-2.13, so most recent binaries will probably not work out of the box. We will probably need a statically linked ARM distribution with the embedded ZeroConf library. I have tried building it using OpenWRT and extracting the .ipk files, but the unit refused the binaries.

If you have some pointers on compiling such a binary, it would be much appreciated.

The next step for me is to try Buildroot as suggested by @joerg-krause , but to be honest I fear this cross compiler business may be a bit too advanced for me.

**Comment by joerg-krause** [comment]
@kidahl The binary I send you was build with Buildroot. Here is the defconfig I use:

```
BR2_arm=y
BR2_cortex_a9=y
BR2_ARM_ENABLE_NEON=y
BR2_ARM_ENABLE_VFP=y
BR2_ARM_EABI=y
BR2_ARM_FPU_NEON=y
BR2_STATIC_LIBS=y
BR2_TOOLCHAIN_EXTERNAL=y
BR2_INIT_NONE=y
BR2_SYSTEM_BIN_SH_NONE=y
# BR2_PACKAGE_BUSYBOX is not set
BR2_PACKAGE_OPENSSL=y
BR2_PACKAGE_SHAIRPORT_SYNC=y
BR2_PACKAGE_SHAIRPORT_SYNC_LIBSOXR=y
# BR2_TARGET_ROOTFS_TAR is not set
```

You can easily reproduce the build by coping this file as let's say `one_defconfig` into the configs directory inside buildroot. `make menuconfig` will show you the options menu. Rebuild everything with `make clean all`.

**Comment by kidahl** [comment]
Excellent, Buildroot it is. As for the compilation of Shairport-sync itself, did you make modifications to enable static compilation? What other options did you enable (soxr, avahi etc)?

FYI, now is probably the time to get a great price on one of the Olive Ones for sale on eBay ;-)

**Comment by joerg-krause** [comment]
No, Buildroot takes care of the static build. Soxr and OpenSSL are enabled. Avahi is not as it has some issues in a static build context, instead the embedded tinysvcmdns is used.

**Comment by mikebrady** [comment]
Did @joerg-krause's binary actually work?

BTW the only Olive One I can see on eBay is €566!

**Comment by kidahl** [comment]
Yes it did, that was a fairly impressive shot in the dark @joerg-krause  :-)

The unit is an audipohile-grade amp, there is a reason why we backers did´t just buy a RasPi instead. But 566EUR was not cheap given the current lack of official support.. If you are still interested I can broadcast it on our Slack channel. I know there has been some announcements. Maybe Jörg just ruined that chance ;)

**Comment by mikebrady** [comment]
That is verrrry impressive work, @joerg-krause! Well done.

**Comment by joerg-krause** [comment]
You are welcome!

The binary is build with metadata support.

@kidahl You wrote about some crashes in the mail. Can you post a log?

**Comment by kidahl** [comment]
The shell just says "Segmentation Fault". I can not for the life of me find a relevant log on this system. There are no logs where /etc/syslog.conf is configured to write kern.* messages.

**Comment by kidahl** [comment]
I have found a way to enable error logging, I will post here if it crashes again..

**Comment by kidahl** [comment]
@joerg-krause I finally got something useful out of the segmentation fault.
The system does not logg anything at att: no logs, nothing in dmesg, no syslog config changes have any effect (and yes, I know that Busybox has a custom syslogd).
I am also unable to create log entries by using logger to test. 

But I was able to capture a coredump after setting ulinit. I have tried getting info out of this file by using gdb on OSX without success, and the unit itself does not have an embedded gdb. 

However, I hope you can make something out of the dump.

My impression is that the process almost always segfaults when I change client, or if the current client has "timed out" and I do a reconnect.
[coredump.zip](https://github.com/mikebrady/shairport-sync/files/440597/coredump.zip)

Thanks,
Karl Ivar

**Comment by mikebrady** [comment]
Fixed – see #388
